lib/src/api/opt/auth.rs

```diff
@@ -12,6 +13,108 @@ pub struct Signup;
 #[derive(Debug)]
 pub struct Signin;
 
+/// Credentials level
+#[derive(Debug, Clone)]
+pub enum CredentialsLevel {
+	Root,
+	Namespace,
+	Database,
+	Scope,
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum Error {
+	#[error("Username is needed for authentication but it was not provided")]
+	NoUsername,
+	#[error("Password is needed for authentication but it was not provided")]
+	NoPassword,
+	#[error("Namespace is needed for authentication but it was not provided")]
+	NoNamespace,
+	#[error("Database is needed for authentication but it was not provided")]
+	NoDatabase,
+	#[error("Scope is needed for authentication but it was not provided")]
+	NoScope,
+	#[error("Scope params are needed for authentication but they were not provided")]
+	NoScopeParams,
+}
+
+/// Construct a Credentials instance for the given auth level
+#[derive(Debug, Default)]
+pub struct CredentialsBuilder<'a> {
+	/// The auth username
+	pub username: Option<&'a str>,
+	/// The auth password
+	pub password: Option<&'a str>,
+	/// The auth namespace
+	pub namespace: Option<&'a str>,
+	/// The auth database
+	pub database: Option<&'a str>,
+	/// The auth scope
+	pub scope: Option<&'a str>,
+}
+
+impl<'a> CredentialsBuilder<'a> {
+	// Builder methods
+	pub fn with_username(mut self, username: Option<&'a str>) -> Self {
```

Comment:
I think `impl Into<Option<&'a str>>` would be better here so we can just do `with_username("username")` instead of `with_username(Some("username"))`. The same applies to the other methods below.